pkg/shared/utils/sync/distributed_lock_mgr.go

```diff
@@ -0,0 +1,69 @@
+package sync
+
+import (
+	"fmt"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/shared/utils/arrays"
+	"gorm.io/gorm"
+	"sync"
+)
+
+var mgrLock sync.Mutex
+
+var _ DistributedLockMgr = &distributedLockMgr{}
+
+func NewDistributedLockMgr(db *gorm.DB) DistributedLockMgr {
+	return &distributedLockMgr{
+		db:           db,
+		locks:        map[string][]DistributedLock{},
```

Comment:
Why is there a list of DistributedLock for each LockID?

Reply:
Each thread must use a distinct `DistributedLock` to hold a separate transaction that waits for the lock’s release. Using a single `DistributedLock` object would result in an `Unlock` call unblocking all threads.

Reply:
> Each thread must use a distinct `DistributedLock` to hold a separate transaction that waits for the lock’s release. Using a single `DistributedLock` object would result in an `Unlock` call unblocking all threads.

I multiple threads are locked then when unlock is called shouldn't all the other threads be unlocked?

Reply:
No. We don't have ONE thread performing the lock and a number of threads waiting for it to be released.
We have a number of threads trying to perform the lock.
Suppose we have 3 threads TH1, TH2, TH3:
```
- instant T1: TH1, TH2 and TH3 are calling the Lock method
- instant T2: TH1 obtains the lock. TH2 and TH3 are blocked, waiting to acquire the lock
- instant T3 : TH1 releases the lock. TH2 is now able to perform the lock. TH3 now waits for TH2 to release the lock
- instant T4: TH2 releases the lock. TH3 can now acquire the lock.
```

Reply:
@ziccardi I still struggling to see why we need to track the lists. Maybe, I am missing something.

I see here that we find only one lock that's current locked: https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/c2559b7bc0edcda2f98475a841a98af496dd2666/pkg/shared/utils/sync/distributed_lock_mgr.go#L56 can't we just track this instead of keeping track of all the lists?

Reply:
Nope, because in that case when the `unlock` removes the object from the `map` it would remove the lock for all the threads that are currently waiting on that thread.
Moreover, each `DistributedLock` object holds a reference to its transaction.

Reply:
Thanks @ziccardi I'll share here what's been discussed as potential future improvement;

We tried tracking only the lock that's acquired but the integration tests where failing. We aren't sure where exactly the race condition was coming from to cause the failure: was it test related or implementation related, that's what we didn't get to nailing down and we decided to defer that improvement and consider it a future improvement in a separate followup PR.